### PR TITLE
Add new li when pressing enter at end of existing li

### PR DESCRIPTION
**Pull request by cmc333333**
If the cursor's at the end of the text of an existing li (and there are no child nodes, like footnotes), pressing Enter will create a new li immediately following the current, renumber the lis, and place the cursor in the new one.

Renumbering inspects the first li and uses that as a template for all following (meaning we can have arbitrary text in the marker). It'd be a good idea to call the renumbering operating when deleting lis and changing indents, but this doesn't implement those scenarios.

## Looks Like
add-bullet
add-ordered
renumbering

### CODE REVIEW

File: api/tslint.yaml [hunk 1]
```diff
@@ -7,3 +7,5 @@ rules:
   # Deprecated in favor of no-this-assignment
   no-var-self: false
   no-this-assignment: true
+  # Removing this as we don't want to have to mix str and literal keys
+  object-literal-key-quotes: false
```
Comment (toolness) [inline api/tslint.yaml]: Ah nice, thanks for adding this! 🎉

File: api/package.json [hunk 2]
```diff
@@ -43,6 +43,8 @@
     "prosemirror-menu": "^1.0.0",
     "prosemirror-state": "^1.0.0",
     "prosemirror-view": "^1.0.5",
+    "repeat-string": "^1.6.1",
+    "romanize": "^1.1.1",
```
Comment (toolness) [inline api/package.json]: Sometimes I wish JSON files supported comments, so we could add a little bit of context briefly explaining what a module does and why we need it.

Also, `repeat-string` makes me wish JS supported `"some string" * 5` like Python does.

File: api/document/js/list-utils.ts [hunk 3]
```diff
@@ -1,29 +1,83 @@
...
   }
   return defaultBullet;
 }
+
+function makeIntToLetter(initial: string) {
+  const offset = initial.charCodeAt(0);
+
+  return (idx: number) => {
+    const count = Math.floor(idx / 26) + 1;
+    const letter = String.fromCharCode(offset + idx % 26);
```
Comment (toolness) [inline api/document/js/list-utils.ts]: I think the meaning of `26` is fairly obvious from context here, though ordinarily I'd suggest making it a named constant like `NUM_LETTERS_IN_ALPHABET`.  Not sure if this would actually be helpful in this case or not.

File: api/document/js/commands.ts [hunk 4]
```diff
@@ -69,3 +70,46 @@ export function makeSaveThenXml(api: Api) {
...
   };
 }
+
+const inLi = (pos: ResolvedPos) => (
+  pos.depth >= 3
+  && pos.node(pos.depth).type === schema.nodes.inline
+  && pos.node(pos.depth - 1).type === schema.nodes.para
+  && pos.node(pos.depth - 2).type === schema.nodes.listitem
+  && pos.node(pos.depth - 3).type === schema.nodes.list
+);
```
Comment (toolness) [inline api/document/js/commands.ts]: This makes me wish we had a sort of "CSS selectors, but for ProseMirror document schema".

File: api/document/js/commands.ts [hunk 5]
```diff
@@ -69,3 +70,46 @@ export function makeSaveThenXml(api: Api) {
...
+const atEndOfLi = (pos: ResolvedPos) => (
+  pos.depth >= 2
+  && pos.pos === pos.end(pos.depth)
+  && pos.pos === pos.end(pos.depth - 1) - 1
+  && pos.pos === pos.end(pos.depth - 2) - 2
+);
+
+export function addListItem(state, dispatch?) {
```
Comment (toolness) [inline api/document/js/commands.ts]: It might be useful, both for documentation and type-checking goodness, to annotate `state` as an [`EditorState`](https://prosemirror.net/docs/ref/#state.EditorState) here (and in other function declarations that take `state`).

It _might_ also be useful to define `dispatch` as `((tr: Transaction) => void) | null` (or make a type alias, since lots of functions take it) but up to you really.

File: api/document/js/commands.ts (hunk 4 above)
Comment (cmc333333) [reply api/document/js/commands.ts]: Agreed. The `pathToResolvedPos()` function implements some of that, but it's pretty rudimentary.

File: api/document/js/commands.ts (hunk 5 above)
Comment (cmc333333) [reply api/document/js/commands.ts]: Good call, can do.

File: api/document/js/__tests__/commands.test.ts [hunk 6]
```diff
@@ -187,9 +191,59 @@ describe('makeSaveThenXml()', () => {
...
+  }
+
+  it('requires the cursor be in the right position', () => {
+    const inIntro = makeState(['para', 'inline', 'int'.length]);
+    const endIntro = makeState(['para', 'inline', 'intro'.length]);
+    const middleOfLi = makeState(['list', 'listitem', 'para', 'inline', 'a'.length]);
+    const endOfFirstPara = makeState(
+      ['list', new NthType(1, 'listitem'), 'para', 'inline', 'bbb first'.length]);
+    [inIntro, endIntro, middleOfLi, endOfFirstPara].forEach(state =>
+      expect(addListItem(state)).toBe(false));
```
Comment (toolness) [inline api/document/js/__tests__/commands.test.ts]: This reminded me of py.test's concept of "parameterized tests" and after a bit of googling it turned up [`jest-each`](https://www.npmjs.com/package/jest-each).  If we think we'll be writing tests like this often, it might be worth adding later, outside of this PR.

File: api/document/js/__tests__/commands.test.ts [hunk 7]
```diff
@@ -187,9 +191,59 @@ describe('makeSaveThenXml()', () => {
...
+    const init = makeState(['list', 'listitem', 'para', 'inline', 'aaa'.length]);
+    const result = executeTransform(init, addListItem);
+    expect(result.selection.anchor).toBe(pathToResolvedPos(
+      result.doc,
+      ['list', new NthType(1, 'listitem'), 'para', 'inline'],
+    ).pos);
+    expect(result.selection.head).toBe(pathToResolvedPos(
+      result.doc,
+      ['list', new NthType(1, 'listitem'), 'para', 'inline', ' '.length],
+    ).pos);
```
Comment (toolness) [inline api/document/js/__tests__/commands.test.ts]: I wonder if it'd be more readable to write this as something like:

```js
const anchorPath = ['list', new NthType(1, 'listitem'), 'para', 'inline'];
expect(result.selection.anchor)
  .toBe(pathToResolvedPos(result.doc, anchorPath).pos);
expect(result.selection.head)
  .toBe(pathToResolvedPos(result.doc, [...anchorPath, ' '.length]).pos);
```

I dunno.

File: api/document/js/list-utils.ts (hunk 3 above)
Comment (cmc333333) [reply api/document/js/list-utils.ts]: I'll play with it a bit to see if there's something helpful and concise. I'll also add a test around the boundary.

File: api/document/js/commands.ts [hunk 8]
```diff
@@ -69,3 +70,46 @@ export function makeSaveThenXml(api: Api) {
...
+  const pos = state.selection.$head;
+  if (!inLi(pos) || !atEndOfLi(pos)) {
+    return false;
+  }
+  if (!dispatch) {
+    return true;
+  }
+
+  const endOfLi: number = pos.end(pos.depth - 2);
```
[the remark this answers is not in the document]
Comment (cmc333333) [reply api/document/js/commands.ts]: I'll take another stab with more comments and perhaps more consts. I agree that ProseMirror's `ResolvedPos` isn't super descriptive with its method names. We may want to extend it.

File: api/document/js/__tests__/commands.test.ts [hunk 9]
```diff
@@ -2,19 +2,22 @@ jest.mock('../Api');
...
   appendNearBlock,
   appendParagraphNear,
   makeSave,
   makeSaveThenXml,
 } from '../commands';
+import { collectMarkers } from '../list-utils';
 import schema, { factory } from '../schema';
 import serializeDoc from '../serialize-doc';
-import pathToResolvedPos, { NthType } from '../path-to-resolved-pos';
+import pathToResolvedPos, { NthType, SelectionPath } from '../path-to-resolved-pos';
```
[the remark this answers is not in the document]
Comment (cmc333333) [reply api/document/js/__tests__/commands.test.ts]: :+1: I'll add a follow up PR for that

File: api/document/js/__tests__/commands.test.ts (hunk 6 above)
Comment (cmc333333) [reply api/document/js/__tests__/commands.test.ts]: Good idea; I'll open a ticket.

File: api/document/js/__tests__/commands.test.ts (hunk 7 above)
Comment (cmc333333) [reply api/document/js/__tests__/commands.test.ts]: I agree, will implement.

**Comment by cmc333333** [comment]
Thanks @toolness! Updated to address; I'll now rebase and merge.
